modyn/storage/src/internal/file_watcher/file_watcher_watchdog.cpp

```diff
@@ -0,0 +1,162 @@
+ *
+ * Also remove the FileWatcher thread from the map of FileWatcher threads
+ *
+ * @param dataset_id The id of the dataset to start a FileWatcher thread for
+ */
+void FileWatcherWatchdog::stop_file_watcher_thread(int64_t dataset_id) {
+  if (file_watcher_threads_.contains(dataset_id)) {
+    // Set the stop flag for the FileWatcher thread
+    file_watcher_thread_stop_flags_[dataset_id].store(true);
+    // Wait for the FileWatcher thread to stop
+    if (file_watcher_threads_[dataset_id].joinable()) {
+      file_watcher_threads_[dataset_id].join();
+    }
+    auto file_watcher_thread_it = file_watcher_threads_.find(dataset_id);
+    if (file_watcher_thread_it == file_watcher_threads_.end()) {
+      SPDLOG_ERROR("FileWatcher thread for dataset {} not found", dataset_id);
+    } else {
+      file_watcher_threads_.erase(file_watcher_thread_it);
+    }
+
+    auto file_watcher_dataset_retries_it = file_watcher_dataset_retries_.find(dataset_id);
+    if (file_watcher_dataset_retries_it == file_watcher_dataset_retries_.end()) {
+      SPDLOG_ERROR("FileWatcher thread retries for dataset {} not found", dataset_id);
+    } else {
+      file_watcher_dataset_retries_.erase(file_watcher_dataset_retries_it);
+    }
+
+    auto file_watcher_thread_stop_flags_it = file_watcher_thread_stop_flags_.find(dataset_id);
+    if (file_watcher_thread_stop_flags_it == file_watcher_thread_stop_flags_.end()) {
+      SPDLOG_ERROR("FileWatcher thread stop flag for dataset {} not found", dataset_id);
+    } else {
+      file_watcher_thread_stop_flags_.erase(file_watcher_thread_stop_flags_it);
+    }
+  } else {
+    SPDLOG_ERROR("FileWatcher thread for dataset {} not found", dataset_id);
+  }
+}
+
+/*
+ * Watch the FileWatcher threads and start/stop them as needed
+ */
+void FileWatcherWatchdog::watch_file_watcher_threads() {
+  soci::session session = storage_database_connection_.get_session();
+
+  int64_t number_of_datasets = 0;
+  session << "SELECT COUNT(dataset_id) FROM datasets", soci::into(number_of_datasets);
+
+  if (number_of_datasets == 0) {
+    if (file_watcher_threads_.empty()) {
+      // There are no FileWatcher threads running, nothing to do
+      return;
+    }
+    // There are no datasets in the database, stop all FileWatcher threads
+    for (auto& file_watcher_thread_flag : file_watcher_thread_stop_flags_) {
+      file_watcher_thread_flag.second.store(true);
+    }
+    for (auto& file_watcher_thread : file_watcher_threads_) {
+      if (file_watcher_thread.second.joinable()) {
+        file_watcher_thread.second.join();
+      }
+    }
+    file_watcher_threads_.clear();
+    file_watcher_dataset_retries_.clear();
+    file_watcher_thread_stop_flags_.clear();
+    return;
+  }
+
+  std::vector<int64_t> dataset_ids(number_of_datasets);
+  session << "SELECT dataset_id FROM datasets", soci::into(dataset_ids);
+
+  const std::vector<int64_t> running_file_watcher_threads = get_running_file_watcher_threads();
```

Comment:
Soci can't handle an unordered_set:

```
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/exchange-traits.h:41:16: error: no member named 'x_type' in 'soci::details::exchange_traits<std::unordered_set<long long>>'
            >::x_type
            ~~~^
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/bind-values.h:211:48: note: in instantiation of template class 'soci::details::exchange_traits<std::unordered_set<long long>>' requested here
    { exchange(do_into(ic.t, typename details::exchange_traits<T>::type_family())); }
                                               ^
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/bind-values.h:143:9: note: in instantiation of function template specialization 'soci::details::into_type_vector::exchange_<std::unordered_set<long long>>' requested here
        exchange_(ic, (typename boost::fusion::traits::is_sequence<T>::type *)NULL);
        ^
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/statement.h:50:14: note: in instantiation of function template specialization 'soci::details::into_type_vector::exchange<std::unordered_set<long long>, void>' requested here
    { intos_.exchange(ic); }
             ^
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/statement.h:199:78: note: in instantiation of function template specialization 'soci::details::statement_impl::exchange<std::unordered_set<long long>, void>' requested here
    void exchange(details::into_container<T, Indicator>  const &ic) { impl_->exchange(ic); }
                                                                             ^
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/ref-counted-statement.h:92:31: note: in instantiation of function template specialization 'soci::statement::exchange<std::unordered_set<long long>, void>' requested here
    void exchange(T &t) { st_.exchange(t); }
                              ^
/Users/viktorgsteiger/Documents/modyn/build/_deps/soci-src/src/core/../../include/soci/once-temp-type.h:49:16: note: in instantiation of function template specialization 'soci::details::ref_counted_statement::exchange<const soci::details::into_container<std::unordered_set<long long>, void>>' requested here
        rcst_->exchange(ic);
               ^
/Users/viktorgsteiger/Documents/modyn/modyn/storage/src/internal/file_watcher/file_watcher_watchdog.cpp:112:47: note: in instantiation of function template specialization 'soci::details::once_temp_type::operator,<std::unordered_set<long long>, void>' requested here
  session << "SELECT dataset_id FROM datasets", soci::into(dataset_ids);
                                              ^
```

Reply:
Okay, I would still insert into a vector via soci, then do vector -> set, and then use the set to avoid find if. Should still be cheaper (probably) :D